File: CodingInterview/CodingInterview/Arrays/Longest_Common_Prefix.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CodingInterview.Arrays
{
    class Longest_Common_Prefix
    {
        public static string LongestCommonPrefix(string[] strs)
        {
            /*
             S= O(m) //var result changes after every loop
            T= O(mn) //two nested loops for different collections
             */
            string result = "";

            if (strs.Length == 0) return result;

            for (int i = 0; i < strs[0].Length; i++) //m
            {
                for (int j = 1; j < strs.Length; j++) //n
                {
                    if (i >= strs[j].Length || strs[0][i] != strs[j][i])
                    {
                        return result;
                    }
                }
                result += strs[0][i];
            }

            return result;
        }
    }
}
/*
 14. Longest Common Prefix
Easy

Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"
Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
 
Constraints:

0 <= strs.length <= 200
0 <= strs[i].length <= 200
strs[i] consists of only lower-case English letters.
 */

File: CodingInterview/CodingInterview/Arrays/Longest_Common_Prefix.js
/**
 * 14. Longest Common Prefix - Leet Code
 * 
 Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

Example 1:

Input: ["flower","flow","flight"]
Output: "fl"
Example 2:

Input: ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
Note:

All given inputs are in lowercase letters a-z.
 */

/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function (strs) {
    let commonPrefix = '';

    if (strs === null || strs.length === 0)
        return commonPrefix;

    let idx = 0;
    for (let c of strs[0]) {

        for (let str = 1; str < strs.length; str++) {
            if (idx >= strs[str].length || c !== strs[str][idx]) {
                return commonPrefix;
            }
        }

        commonPrefix += c;
        idx++;
    }

    return commonPrefix;
};